src/generic/LaurentPoly.jl

```diff
@@ -0,0 +1,265 @@
+###############################################################################
+#
+#   LaurentPoly.jl : Generic Laurent polynomials over rings
+#
+###############################################################################
+
+import AbstractAlgebra: monomials_degrees
+using AbstractAlgebra: monomial_degree, degrees_range
+
+###############################################################################
+#
+#   Data type and parent object methods
+#
+###############################################################################
+
+parent_type(::Type{LaurentPolyWrap{T, PE}}) where {T, PE} =
+   LaurentPolyWrapRing{T, parent_type(PE)}
+
+elem_type(::Type{LaurentPolyWrapRing{T, PR}}) where {T, PR} =
+   LaurentPolyWrap{T, elem_type(PR)}
+
+parent(p::LaurentPolyWrap) = LaurentPolyWrapRing(parent(p.poly))
+
+base_ring(R::LaurentPolyWrapRing) = base_ring(R.polyring)
+
+var(R::LaurentPolyWrapRing) = var(R.polyring)
+
+symbols(R::LaurentPolyWrapRing) = symbols(R.polyring)
+
+nvars(R::LaurentPolyWrapRing) = nvars(R.polyring)
+
+characteristic(R::LaurentPolyWrapRing) = characteristic(R.polyring)
+
+
+###############################################################################
+#
+#   Basic manipulation
+#
+###############################################################################
+
+monomials_degrees(p::LaurentPolyWrap) = p.mindeg .+ (0:degree(p.poly))
+
+coeff(p::LaurentPolyWrap, i::Int) =
+   i < p.mindeg ? zero(base_ring(p)) : coeff(p.poly, i - p.mindeg)
+
+function _enable_deg!(p::LaurentPolyWrap, i::Int)
+   diff = p.mindeg - i
+   if diff > 0
+      p.mindeg = i
+      p.poly = shift_left(p.poly, diff)
+   end
+   nothing
+end
+
+# the underlying storage is adjusted (increased) to allow setting the coeff
+function setcoeff!(p::LaurentPolyWrap, i::Int, a)
+   _enable_deg!(p, i)
+   setcoeff!(p.poly, i - p.mindeg, a)
+   p
+end
+
+iszero(p::LaurentPolyWrap) = iszero(p.poly)
+
+isone(p::LaurentPolyWrap) = ismonomial(p, 0, rec=false)
+
+zero(R::LaurentPolyWrapRing) = LaurentPolyWrap(zero(R.polyring))
+one(R::LaurentPolyWrapRing) = LaurentPolyWrap(one(R.polyring))
+
+gen(R::LaurentPolyWrapRing) = LaurentPolyWrap(gen(R.polyring))
+
+isgen(p::LaurentPolyWrap) = ismonomial(p, 1, rec=false)
+
+# only an optimization over the default Base implementation (maybe 1.4 speed-up)
+deepcopy_internal(p::LaurentPolyWrap, dict::IdDict) =
+   LaurentPolyWrap(deepcopy_internal(p.poly, dict), p.mindeg)
+
+
+###############################################################################
+#
+#   Unary and Binary operations
+#
+###############################################################################
+
+-(p::LaurentPolyWrap) = LaurentPolyWrap(-p.poly, p.mindeg)
+
+function +(p::LaurentPolyWrap, q::LaurentPolyWrap)
+   if p.mindeg > q.mindeg
+      p, q = q, p
+   end
+   p_, q_ = p.poly, q.poly
+   if p.mindeg < q.mindeg
+      q_ = shift_left(q_, q.mindeg - p.mindeg)
+   end
+   LaurentPolyWrap(p_ + q_, p.mindeg)
```

Comment:
There is an implicit decision here (and elsewhere) to not store the laurent polynomial in a "canonical" form (i.e. there is no attempt choose the maximal possible value for `mindeg`). Which I guess might make sense for performance reasons. But in any case, I would recommend to document this explicitly; say in a source code comment next to the definition of `struct LaurentPolyWrap`, where one could explain the meaning of the members `poly` and `mindeg` and that they are not uniquely determined.

Reply:
Yes good point, I will add these comments. I didn't determine with benchmarks that not canonicalising is generally more efficient; but it at least seemed simpler for an initial implementation.